scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+(ns cook.kubernetes.api
+  (:require [clojure.tools.logging :as log]
+            [plumbing.core :as pc])
+  (:import
+    (com.google.gson JsonSyntaxException)
+    (com.twosigma.cook.kubernetes WatchHelper)
+    (io.kubernetes.client ApiClient ApiException)
+    (io.kubernetes.client.apis CoreV1Api)
+    (io.kubernetes.client.custom Quantity Quantity$Format)
+    (io.kubernetes.client.models V1Pod V1Container V1Node V1Pod V1Container V1ResourceRequirements V1EnvVar
+                                 V1ObjectMeta V1PodSpec V1PodStatus V1ContainerState V1DeleteOptionsBuilder
+                                 V1DeleteOptions)
+    (io.kubernetes.client.util Watch)
+    (java.util.concurrent Executors ExecutorService)))
+
+
+(def ^ExecutorService kubernetes-executor (Executors/newFixedThreadPool 2))
+
+(defn handle-watch-updates
+  [state-atom ^Watch watch key-fn callback]
+  (while (.hasNext watch)
+    (let [update (.next watch)
+          item (.-object update)
+          prev-item (get @state-atom (key-fn item))]
+      (case (.-type update)
+        "ADDED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
+        "MODIFIED" (swap! state-atom (fn [m] (assoc m (key-fn item) item)))
+        "DELETED" (swap! state-atom (fn [m] (dissoc m (key-fn item)))))
+      (when callback
+        (try
+          (callback prev-item item)
+          (catch Exception e
+            (log/error e "Error while processing callback")))))))
```

Comment:
Should we also log `prev-item` and `item` here?

Reply:
They can be noisy, but probably good to at least have something like `(when item (key-fn item))`